feat(typed_state): add TypedState util class

Thibault-Pelletier opened the pull request:
- Add state to dataclass state proxies to simplify binding structured and strongly typed data with the trame state while keeping the reactivity granularity for leaf components.

jourdain commented:
Great first step, but I feel we can do a better job in our exposed API. We should see what API/usage we aim to enable and then works toward it. I think you have all the internal pieces... Now we need to figure out how to make it super easy to use.

Just one usage idea

```python
class App(TypedTrameApp):
   def __init__(self, server=None):
       super().__init__(server, typed_state=YourTypedStateClass)

    @change("a")
    def on_change(a, **_):
          assert a == self.typed_state.a
          print(self.typed_state.a)
```

Thibault-Pelletier commented:
Thanks for the feedback!

The current implementation can already work as is for individual components IMO.
One of the driving idea behind this PR was for apps with multiple widgets implements in Python.
If these widgets are composed of multiple UI elements, then instantiating them multiple times and handling there state becomes hard to do manually and you end up using string formatting to handle the different UI components individually.

See an [example of that in trame-slicer here](https://github.com/KitwareMedical/trame-slicer/blob/main/trame_slicer/views/trame_helper.py#L23).

Having a higher level API for the TrameApp component would probably ease usage for new users, but I'm not sure how to address the change decorator API.

The current implementation can use namespace prefixes for the state and automatically [adds in the name of the source class to the state id](https://github.com/Kitware/trame-server/pull/52/files#diff-e0b7ce4522d785763015110a7ea8df05d7c9d9479367d690018b5c35615942c8R264). This makes the binding to a simple string less straightforward. 

On my end, I almost never use raw strings or kwargs for the state change connection as I always break them when refactoring code and I prefer using simple descriptor patterns and structs instead.

This change connection could work using this strategy by defining the API inside the init like follows : 

```python
class App(TypedTrameApp):
   def __init__(self, server=None):
       super().__init__(server, typed_state=YourTypedStateClass, state_encoder=<>, state_namespace=<> )

      @self.change(self.state_id.inner.a_value)
      def on_change(a):   # API change to variadic arg with same size as change decorator disconnected from actual kwarg
            assert a == self.typed_state_name.inner.a_value
            print(self.typed_state_name.inner.a_value)
``` 

Looking forward to your thoughts!

jourdain commented:
I guess for the decorator path we could have some kind of self resolver. Something like below. Either way, I don't want to create closure methods just to get access to the self at the method definition.

```python
class Something(TrameApp):

    @change(LocalResolver("state_id", "inner.a_value"))
    def on_change(**_):
       ...
```

Unless you are going to run into similar string / refactoring issue?

In which case, we could envision some mapping helper to use in the constructor like.

```python
def __init__(self, ...):
    self.bind_changes({
         self.state_id.inner.value_a: self.on_change,
         self.state_id.outer.value_b: self.on_other_change,
    })
```

jourdain commented:
Also I would rather use a decorator pattern instead of a method execution to get the data and names.

Instead of doing 

```python
data, data_names = state_proxies(
    TrameState, state, state_encoder=CollectionEncoderDecoder([CustomEncoder()])
)
```

doing something like

```
def __init__(...):
    # ...
    self.type_state = TypedState(self.state, TrameState, state_encoder=...)

    self.type_state.data.inner.text_value
    self.type_state.name.inner.text_value
```

What do you think?

Thibault-Pelletier commented:
I just added some code to implement the TypedState class with some overloads for the bind_change method.
Happy to discuss / implement further API sugar :)

jourdain commented:
I would not change the class decorator, but you can extend the [TrameApp base class](https://github.com/Kitware/trame-common/blob/main/src/trame_common/obj/app.py#L7).

jourdain commented on trame_server/utils/state_proxy.py:
```diff
@@ -0,0 +1,420 @@
+    def bind_change(self, change_dict: dict[Any | list[Any], Callable]) -> None: ...
+
+    def bind_change(self, *args) -> None:
+        if len(args) == 1 and isinstance(args[0], dict):
```
you may want to add some docstring to explain usage especially since you have a couple of way to call it.

Thibault-Pelletier replied:
Yes, sorry about that, I pushed a first API example to keep discussing implementation and API here but its missing docstrings.

Let me know how / where I should document the proxy mechanism as well.
I will add the necessary documentation once we agree on how it should work.

Thibault-Pelletier commented on trame_server/utils/state_proxy.py:
```diff
@@ -0,0 +1,420 @@
+    @state.change(*keys)
+    def _on_state_change(**_):
+        values = [state_id_to_field_dict[k].get_value() for k in keys]
+        callback(*values)
```
The default state.change would not work well for typed states for two reasons : 

1. The actual state key used in the state is not easy to guess by a user
2. The call will not use strong types but the State internal storage which for typed state is after its encoder


A bit of detail on the first point : 

To handle an automatic serialization / deserialization of the dataclass to / from trame state and avoid manglings between different dataclass with same names but different classes, I added the name of the src dataclass and the name of the fields into the underlying trame state.

So a user wouldn't be able to infer which arbitrary name was mapped to a given proxy. Since trame change callback expand the dict kwargs, trying to use a user defined name will not work.

```python
@dataclass
class MyData:
    a: int = 1
    b: int = 2


@dataclass
class MyBiggerData:
    my_other_data: MyData = field(default_factory=MyData)
    c: float = 42.0


typed_state = TypedState(state, MyBiggerData)
print(typed_state.name.my_other_data.a)
>>> MyBiggerData__my_other_data__MyData__a

```

On the second point, since the proxy can encode / decode to simplify trame state serialization, there is a need for the proxy to convert whatever trame state is passed for users to get there typed values in the callback.

Expanding on that (implementation missing but coming soon) it becomes possible with the proxy mecanism to also bind structured dataclass to the callback and not just individual fields.

The callback will be called for any individual changes to the dataclass and users can manipulate the structured object directly.  

On this note, I'm not sure if it makes more sense for users to manipulate the data or the proxy for this callback. 
If we pass the dataclass directly, then users can manipulate it in natural dataclass ways. Mutating the input would not impact the state. If we pass the state proxy, mutating it would impact the state but you would need to use `as_dataclass` method to maniuplate the real dataclass instance...

Example of this below : 

```python
@dataclass
class ClockData:
   t: time
   d: date

@dataclass
class MyState:
   clock_data:  ClockData


def on_clock_change(clock: ClockData):
   assert isinstance(clock.t, time) # Converted strong time type
   assert isinstance(clock.d, date) # Converted strong date type
   
typed_state = TypedState(state, my_state)
typed_state.bind_change(typed_state.name.clock_data, on_clock_change)
```

jourdain replied:
I think I like it...

Thibault-Pelletier commented:
> I would not change the class decorator, but you can extend the [TrameApp base class](https://github.com/Kitware/trame-common/blob/main/src/trame_common/obj/app.py#L7).

If the bind_change API is different from the default state one, maybe the TypedState should remain separate from the TrameApp class and decorators.

Users whanting to use TypedState this way sould be able to use multiple inheritence without much problems.

On a related note, should I move the PR to the trame_common repo to make these kind of changes easy if needed?

jourdain commented:
I think if you encapsulate everything inside a TypedState class that you can use as a state decorator, that will be great. For the location, if you don't have any dependency to trame-server, then yes move to trame-common. If not, I'm fine keeping it in trame-server where the state is declared.

So to summarize the API we talked about, it will be as follow (let me know if I missed anything.

```python
typed_state = TypedState(server.state, ...)

typed_state.name.* # get state name for UI
typed_state.data.* # get dataclass instances
typed_state.bind_changes({
    # I would force a single {} arg all the time to prevent usage confusion
    # callback will be called with resolved value as args
   (type_state.name.a, type_state.name.b): callback(a, b),
   ...
})

typed_state.as_dataclass(...) # helper for data conversion?
typed_state.??? # anything else?
```

jourdain commented:
Question, in your example you have

```python
typed_state = TypedState(
    state, TrameState, state_encoder=CollectionEncoderDecoder([CustomEncoder()])
)
```

Could we just have

```python
typed_state = TypedState(
    state, TrameState, state_encoders=[CustomEncoder()])
)
```

Or do you expect people to override the `CollectionEncoderDecoder`?

Thibault-Pelletier commented:
> Question, in your example you have
> 
> ```python
> typed_state = TypedState(
>     state, TrameState, state_encoder=CollectionEncoderDecoder([CustomEncoder()])
> )
> ```
> 
> Could we just have
> 
> ```python
> typed_state = TypedState(
>     state, TrameState, encoders=[CustomEncoder()])
> )
> ```
> 
> Or do you expect people to override the `CollectionEncoderDecoder`?

This was mainly to have only one interface API, but it probably makes more sense to expect a list of encoders if any and always use the collection encoder internally.

I'll make the change!

Thibault-Pelletier commented on examples/dataclass_typed_state.py:
```diff
@@ -0,0 +1,209 @@
+
+from trame.app import get_server
+from trame.ui.vuetify import SinglePageLayout
+from trame.widgets import vuetify2 as vuetify
```
Sounds good! I'll make the change!

Thibault-Pelletier commented:
I just updated the implementation with the following changes : 

- Updated example to use vuetify3
- Changed bind_change API (renamed to bind_changes) to single dict input
- Added a get_dataclass() and set_dataclass method to the typed_state
- Changed the expected encoder to an encoder list and use CollectionEncoder by default under the hood
- Moved all the free functions to TypedState classmethod to remove file clutter
- Update the unit test to only use TypedState as API 

For code location, it probably makes sense to keep it next to the State for now as any changes to the State API will impact the TypedState directly.

Please let me know if you see other things to change!

jourdain commented on trame_server/utils/typed_state.py:
```diff
@@ -0,0 +1,501 @@
+        *,
+        state_namespace="",
+        state_encoders: list[IStateEncoderDecoder] | None = None,
+    ):
```
Small detail, but since you are defining the TypedState class, I would skip the `state_` prefix for __namespace__ and __encoders__.

Thibault-Pelletier replied:
Sounds good, I removed the prefix!

jourdain reviewed (approved):
LGTM, feel free to merge

Thibault-Pelletier commented:
FYI I excluded the `exclude: ^trame_server/LICENSE$` from the `end-of-file-fixer` pre-commit hook. It seems to break LICENSE symlink on Windows...

I don't think I have any admin access to this repo so I'll check with @finetjul if he has access and can launch the pipeline!

finetjul reviewed (approved):
LGTM

finetjul commented:
> Do you need me to merge?

Yes, please. I can't merge.
